### ISSUE TITLE
Optimizer initialization with DDP

### ISSUE THREAD

**Issue by PhilJd** [PyTorchLightning/pytorch-lightning#2902]
#### What is your question?
I would have expected optimizers to always be initialized after parameters have been moved to their destination device.
However, some ddp backends such as
[ddp_backend](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/ddp_backend.py#L174),             [ddp_spawn_backend](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/ddp_spawn_backend.py#L115),            [ddp2_backend](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/ddp2_backend.py#L111)
initialize the optimizer with the CPU parameters before moving the model to the GPU while others such as [gpu_backend](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/gpu_backend.py#L39) pass the GPU parameters.

I'm currently trying to understand two things: 
(1) Where does the linking from the CPU to GPU parameters happen?
(2) Is it actually necessary to initialize the optimizer before moving to the specific device or could it be done afterwards? (Most tutorials initialize the optimizer *after* placing the parameters on the corresponding device, e.g.  [this one](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#basic-use-case))

The reason I'm asking is that I did some parameter/gradient bending to be views into other tensors. This does not work with the current implementation as the optimizer keeps the reference to the CPU parameters with these tweaks but works fine when adapting the pytorch_lightning code by moving the optimizer creation after the model has been moved to the correct device.

**Comment by awaelchli** [comment]
Hi, could you share what you changed to defer the initialization?

**Comment by PhilJd** [comment]
For ddp_backend, I've moved [lines 174-177](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/ddp_backend.py#L174-L177) ( which is mainly `optimizers, lr_schedulers, optimizer_frequencies = self.trainer.init_optimizers(model)`) to line 199, i.e., after `model.cuda(self.trainer.root_gpu)`. Note that so far I have only verified that the parameters receive gradients again, so I don't know if this breaks something else.

**Comment by awaelchli** [comment]
Looks good, I don't see a problem with this. Could you send a PR, then we will see if the tests pass? Then we should also check the other backends to make it consistent.

**Comment by PhilJd** [comment]
Done :)
